ddl/ddl_api.go

```diff
@@ -5172,67 +5162,24 @@ func (d *ddl) AlterIndexVisibility(ctx sessionctx.Context, ident ast.Ident, inde
 	return errors.Trace(err)
 }
 
-func checkPlacementLabelConstraint(label string) (placement.LabelConstraint, error) {
-	r := placement.LabelConstraint{}
-
-	if len(label) < 4 {
-		return r, errors.Errorf("label constraint should be in format '{+|-}key=value', but got '%s'", label)
-	}
-
-	var op placement.LabelConstraintOp
-	switch label[0] {
-	case '+':
-		op = placement.In
-	case '-':
-		op = placement.NotIn
-	default:
-		return r, errors.Errorf("label constraint should be in format '{+|-}key=value', but got '%s'", label)
-	}
-
-	kv := strings.Split(label[1:], "=")
-	if len(kv) != 2 {
-		return r, errors.Errorf("label constraint should be in format '{+|-}key=value', but got '%s'", label)
-	}
-
-	key := strings.TrimSpace(kv[0])
-	if key == "" {
-		return r, errors.Errorf("label constraint should be in format '{+|-}key=value', but got '%s'", label)
-	}
-
-	val := strings.TrimSpace(kv[1])
-	if val == "" {
-		return r, errors.Errorf("label constraint should be in format '{+|-}key=value', but got '%s'", label)
-	}
-
-	r.Key = key
-	r.Op = op
-	r.Values = []string{val}
-	return r, nil
-}
-
-func checkPlacementLabelConstraints(rule *placement.Rule, labels []string) error {
-	for _, str := range labels {
-		label, err := checkPlacementLabelConstraint(strings.TrimSpace(str))
-		if err != nil {
-			return err
-		}
-		rule.LabelConstraints = append(rule.LabelConstraints, label)
-	}
-	return nil
-}
-
-func checkPlacementSpecConstraint(rules []*placement.Rule, rule *placement.Rule, cnstr string) ([]*placement.Rule, error) {
-	cnstr = strings.TrimSpace(cnstr)
+func buildPlacementSpecConstraint(rules []*placement.RuleOp, rule *placement.RuleOp, replicas uint64, cnstr string) ([]*placement.RuleOp, error) {
```

Comment:
It's better to return new rules, instead of appending to old rules in this function.